src/server/controllers/LinkStatisticsController.ts

```diff
@@ -0,0 +1,37 @@
+import { Request, Response } from 'express'
+import { inject, injectable } from 'inversify'
+
+import { DependencyIds } from '../constants'
+import { LinkStatisticsControllerInterface } from './interfaces/LinkStatisticsControllerInterface'
+import { LinkStatisticsServiceInterface } from '../services/interfaces/LinkStatisticsServiceInterface'
+import { UserType } from '../models/user'
+
+@injectable()
+export class LinkStatisticsController
+  implements LinkStatisticsControllerInterface {
+  private linkStatisticsService: LinkStatisticsServiceInterface
+
+  public constructor(
+    @inject(DependencyIds.linkStatisticsService)
+    linkStatisticsService: LinkStatisticsServiceInterface,
+  ) {
+    this.linkStatisticsService = linkStatisticsService
+  }
+
+  public getLinkStatistics: (
+    req: Request,
+    res: Response,
+  ) => Promise<void> = async (req, res) => {
+    const shortUrl = req.query.url
+    if (!shortUrl) res.json(null)
+    const user = req.session?.user as UserType
+    if (!user) res.json(null)
+    const linkStats = await this.linkStatisticsService.getLinkStatistics(
+      user.id,
+      shortUrl,
+    )
+    res.json(linkStats)
```

Comment:
aren't you missing return statements after each call to `res.json`? the server will continue processing even if a response has been returned to the client

Reply:
also get in the habit of using HTTP status codes to indicate the server response e.g. (`res.status(404).send()`), instead of `res.json` which returns an implicit HTTP 200 OK. the frontend should check the HTTP status code and react accordingly.

Reply:
I have modified the code to use return statements and send explicit status codes